lib/routes/session.js

```diff
@@ -87,8 +88,108 @@ module.exports = function (log, db) {
+      method: 'POST',
+      path: '/session/duplicate',
+      config: {
+        auth: {
+          strategy: 'sessionToken'
+        },
+        validate: {
+          payload: {
+            reason: isA.string().max(16).optional()
+          }
+        }
+      },
+      handler: function (request, reply) {
+        log.begin('Session.duplicate', request)
+        const origSessionToken = request.auth.credentials
+
+        return P.resolve()
+          .then(duplicateVerificationState)
+          .spread(createSessionToken)
+          .then(formatResponse)
+          .then(reply, reply)
+
+        function duplicateVerificationState() {
+          // Copy verification state of the token, but generate
+          // independent verification codes.
+          const newVerificationState = [null, null]
+          if (origSessionToken.tokenVerificationId) {
+            newVerificationState[0] = random.hex(origSessionToken.tokenVerificationId.length / 2)
+          }
+          if (origSessionToken.tokenVerificationCode) {
+            newVerificationState[1] = random.hex(origSessionToken.tokenVerificationCode.length / 2)
+          }
+          return newVerificationState
+        }
+
+        function createSessionToken(newVerificationId, newVerificationCode) {
+          // Update UA info based on the requesting device.
+          const {
+            browser: uaBrowser,
+            browserVersion: uaBrowserVersion,
+            os: uaOS,
+            osVersion: uaOSVersion,
+            deviceType: uaDeviceType,
+            formFactor: uaFormFactor
+          } = request.app.ua
+
+          // Copy all other details from the original sessionToken.
+          const sessionTokenOptions = {
+            uid: origSessionToken.uid,
+            // We have to lie a little here and copy the creation time
+            // of the original sessionToken.  Otherwise we would falsely
+            // report a `lastAuthAt` value matching the current timestamp.
+            createdAt: origSessionToken.createdAt,
+            email: origSessionToken.email,
+            emailCode: origSessionToken.emailCode,
+            emailVerified: origSessionToken.emailVerified,
+            verifierSetAt: origSessionToken.verifierSetAt,
+            locale: origSessionToken.locale,
+            mustVerify: !! origSessionToken.mustVerify,
+            tokenVerificationId: newVerificationId,
+            tokenVerificationCode: newVerificationCode,
+            // This prevents the new token from being verified via email code.
+            // That's OK, because we don't send them a new email with the new
+            // verification codes unless they explicitly ask us to resend it,
+            // and resend only handles email links rather than email codes.
+            tokenVerificationCodeExpiresAt: 0,
+            uaBrowser,
+            uaBrowserVersion,
+            uaOS,
+            uaOSVersion,
+            uaDeviceType,
+            uaFormFactor
+          }
```

Comment:
No biggie, but I'm wondering if it might save a (tiny) bit of future maintenance effort here if we leaned on `Object.assign` instead of assigning every property explicitly? Possibly combined with returning an object instead of an array from `duplicateVerificationState` so that those properties can apply using the same keys?